examples/userapi/subdomains.ipynb

```diff
@@ -0,0 +1,954 @@
+    "rho[:,:,int(0.5*shape[2])+1+int(4/h):] = 2.\n",
+    "\n",
+    "origin = (0, 0, 0)\n",
+    "spacing = (h, h, h)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Before defining and creating our subdomains we define the thickness of our absorption layer:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 28,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "nbl = 20 # Number of absorbing boundary layers cells"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now define our subdomains, one for the upper water layer, and one for the rock layers beneath:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 29,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Define our 'upper' and 'lower' SubDomains:\n",
+    "class Upper(SubDomain):\n",
+    "    name = 'upper'\n",
+    "    def define(self, dimensions):\n",
+    "        x, y, z = dimensions\n",
+    "        return {x: x, y: y, z: ('left', int(0.5*shape[2])+1+nbl)}\n",
+    "    \n",
+    "class Lower(SubDomain):\n",
+    "    name = 'lower'\n",
+    "    def define(self, dimensions):\n",
+    "        x, y, z = dimensions\n",
+    "        return {x: x, y: y, z: ('right', shape[2]+nbl-(int(0.5*shape[2])+1))}\n",
+    "\n",
+    "# Create these subdomains:\n",
+    "ur = Upper()\n",
+    "lr = Lower()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Create our model to which we pass the subdomains `ur` and `lr`. Within model they will then be used when creating the model grid:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 30,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "Operator `initdamp` run in 0.04 s\n",
+      "Operator `padfunc` run in 0.01 s\n",
+      "Operator `padfunc` run in 0.01 s\n",
+      "Operator `padfunc` run in 0.01 s\n"
+     ]
+    }
+   ],
+   "source": [
+    "so = 4 # FD space order (Note that the time order is by default 1).\n",
+    "\n",
+    "model = ModelElastic(space_order=so, vp=vp, vs=vs, rho=rho, origin=origin, shape=shape,\n",
+    "                     spacing=spacing, nbl=nbl, subdomains=(ur,lr))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now prepare model parameters along with the functions to be used in the PDE, and the source injection term:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 31,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "s = model.grid.stepping_dim.spacing\n",
+    "\n",
+    "# Source freq. in MHz (note that the source is defined below):\n",
+    "f0 = 0.12\n",
+    "\n",
+    "# Thorbecke's parameter notation\n",
+    "l = model.lam\n",
+    "mu = model.mu\n",
+    "ro = model.irho\n",
+    "\n",
+    "t0, tn = 0., 30.\n",
+    "dt = 0.9*model.critical_dt\n",
```

Comment:
?

Reply:
Needed for numerical stability, meant to mention this in the meeting last week but forgot.

Reply:
Wait what is unstable and when? dt should be fine for any model.

Reply:
If you run with `dt = model.critical_dt` you'll see that the numerics blow up in the denser lower layer (iirc). Will investigate further.

Reply:
see new comment